src/coreclr/crossgen-corelib.proj

```diff
@@ -2,37 +2,40 @@
   <PropertyGroup>
     <!-- NoTargets SDK needs a TFM set. So we used the latest .NETCoreApp supported one. -->
     <TargetFramework>$(NetCoreAppToolCurrent)</TargetFramework>
+
+    <!-- Default for using Crossgen2 when not set externally -->
+    <UseCrossgen2 Condition="'$(UseCrossgen2)' == ''">false</UseCrossgen2>
+
+    <OSPlatformConfig>$(TargetOS).$(TargetArchitecture).$(Configuration)</OSPlatformConfig>
+    <RootBinDir>$([MSBuild]::NormalizeDirectory('$(RepoRoot)', 'artifacts'))</RootBinDir>
+    <LogsDir>$([MSBuild]::NormalizeDirectory('$(RootBinDir)', 'log'))</LogsDir>
+    <BinDir>$([MSBuild]::NormalizeDirectory('$(RootBinDir)', 'bin', 'coreclr', $(OSPlatformConfig)))</BinDir>
+    <IntermediatesDir>$([MSBuild]::NormalizeDirectory('$(RootBinDir)', 'obj', 'coreclr', $(OSPlatformConfig)))</IntermediatesDir>
+    <CrossGenCoreLibLog>$([MSBuild]::NormalizePath('$(LogsDir)', 'CrossgenCoreLib_$(TargetOS)__$(TargetArchitecture)__$(Configuration).log'))</CrossGenCoreLibLog>
+    <ExeExtension Condition="'$(OS)' == 'Windows_NT'">.exe</ExeExtension>
+    <DotNetCli>$([MSBuild]::NormalizePath('$(RepoRoot)', 'dotnet.sh'))</DotNetCli>
+    <DotNetCli Condition="'$(OS)' == 'Windows_NT'">$([MSBuild]::NormalizePath('$(RepoRoot)', 'dotnet.cmd'))</DotNetCli>
+
+    <CoreLibAssemblyName>System.Private.CoreLib</CoreLibAssemblyName>
+    <CoreLibInputPath>$([MSBuild]::NormalizePath('$(BinDir)', 'IL', '$(CoreLibAssemblyName).dll'))</CoreLibInputPath>
+    <CoreLibOutputPath>$([MSBuild]::NormalizePath('$(BinDir)', '$(CoreLibAssemblyName).dll'))</CoreLibOutputPath>
+  </PropertyGroup>
+  <PropertyGroup Condition="'$(BuildArchitecture)' != '$(TargetArchitecture)'">
+    <CrossDir Condition="'$(TargetArchitecture)' == 'arm' or '$(TargetArchitecture)' == 'arm64'">x64</CrossDir>
+    <CrossDir Condition="'$(TargetArchitecture)' == 'arm' and '$(UseCrossgen2)' != 'true' and '$(OS)' == 'Windows_NT'">x86</CrossDir>
+    <CrossDir Condition="'$(TargetArchitecture)' == 'x86' and '$(UseCrossgen2)' == 'true'">$(BuildArchitecture)</CrossDir>
+  </PropertyGroup>
+  <PropertyGroup>
+    <CrossGen1ToolPath>$([MSBuild]::NormalizePath('$(BinDir)', '$(CrossDir)', 'crossgen$(ExeExtension)'))</CrossGen1ToolPath>
+    <CrossGen2ToolPath>$([MSBuild]::NormalizePath('$(BinDir)', '$(CrossDir)', 'crossgen2', 'crossgen2.dll'))</CrossGen2ToolPath>
+    <CrossGenToolPath Condition="'$(UseCrossgen2)' == 'false'">$(CrossGen1ToolPath)</CrossGenToolPath>
+    <CrossGenToolPath Condition="'$(UseCrossgen2)' != 'false'">$(CrossGen2ToolPath)</CrossGenToolPath>
```

Comment:
We should make sure to include some of the ILCompiler.* assemblies in the up-to-date check. That's where most of the crossgen2 logic lives.

Reply:
Do you use auto-generated reference assemblies during build? Unless you do, the exe should always be rebuilt if the dependencies change

Reply:
To be honest I’m not sure if we do in this part of the tree.

Reply:
Confirmed -- this works fine. Ref assembly building is not enabled. In this case, I'd say enabling is a mistake. The primary benefit is not rebuilding the downstream tree, but in this case we almost always want to rebuild the downstream tree.